enhancements/olm-crd-conversion-webhooks.md

```diff
@@ -0,0 +1,99 @@
+---
+title: olm-crd-conversion-webhook-support
+authors:
+  - "@sdhaliwa"
+reviewers:
+  - TBD
+approvers:
+  - TBD
+creation-date: 2020-06-10
+last-updated: 2020-07-05
+status: implementable
+---
+
+# OLM CRD Conversion Webhook Support
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This enhancement covers the necessary steps to add OLM support for [ CRD conversion](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#webhook-conversion) webhooks.
+
+## Definitions
+
+- Webhook: Webhooks are HTTP callbacks, providing a way for notifications to be delivered to an external web server i.e. a simple event-notification via HTTP POST. Examples:  [ CRD conversion](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definition-versioning/#webhook-conversion) webhooks, [validating](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) and [mutating](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#mutatingadmissionwebhook) admission webhooks.
+
+- Conversion Webhook: Webhook that can convert an object from one version to another. 
+
+  When API clients, like kubectl or your controller, request a particular version of your resource, the Kubernetes API server needs to return a result that’s of that version. However, that version might not match the version stored by the API server.
+
+  In that case, the API server needs to know how to convert between the desired version and the stored version. Since the conversions aren’t built in for CRDs, the Kubernetes API server calls out to a webhook to do the conversion instead. 
+
+## Motivation
+
+Many Operators are shipping conversion webhooks to deliver conversion of an object from one version to another. The CustomResourceDefinition API supports a versions field that is used to support multiple versions of custom resources that have been developed. Versions can have different schemas and a conversion webhook is used to convert custom resources between versions.
+
+OLM needs to support operators with crd conversion webhooks and help manage and rotate corresponding certificates.
+
+### Goals
+
+- Operator authors can include crd conversion webhooks in their bundle as a part of a CSV.
+- Operator author will not need to manage [cert](https://www.ssl.com/faqs/what-is-a-certificate-authority/) rotation for webhooks shipped with their operators.
+
+### Non-goals
+
+- This enhancement aims to support AllNamespace/ Global Singleton Operators only.
+- Specifying PodDisruptionBudgets with webhooks is out of scope for this enhancement. 
```

Comment:
I think webhook support without PDBs is an incomplete feature, since it can lead directly to clients getting errors from the k8s API, but I recognize that as an issue beyond this enhancement that affects existing webhook support. Of course API failures can happen, and clients need to be able to operate correctly when they get an error from an API request. But it seems reasonable to maintain uptime for critical-path API services using basic k8s primitives. Hopefully that can be prioritized soon.

Reply:
ok, thanks.